rust/relay/ebpf-turn-router/src/main.rs

```diff
@@ -164,270 +164,1385 @@ fn try_handle_turn_ipv4(ctx: &XdpContext, eth: Eth) -> Result<(), Error> {
+            .add_u16(channel_data_length)
+            .into_checksum();
+
+        if check == 0 {
+            udp.set_check(0xFFFF); // Special case for zero checksum - write 0xFFFF to the wire
+        } else {
+            udp.set_check(check);
         }
+    }
+
+    //
+    // 4. Channel data header
+    //
+    // SAFETY: This is the only mutable instance of `CdHdr` in this scope.
+    let cd = unsafe { ref_mut_at::<CdHdr>(ctx, EthHdr::LEN + Ipv4Hdr::LEN + UdpHdr::LEN)? };
+    cd.number = channel_number.to_be_bytes();
+    cd.length = channel_data_length.to_be_bytes();
+
+    Ok(())
+}
+
+// Convert IPv4 to IPv6 and add channel data
+#[inline(always)]
+fn handle_ipv4_udp_to_ipv6_channel(
+    ctx: &XdpContext,
+    client_and_channel: &ClientAndChannelV6,
+) -> Result<(), Error> {
+    // Expand the packet by 24 bytes for IPv6 header and channel data header
+    const NET_EXPANSION: i32 = -(Ipv6Hdr::LEN as i32 - Ipv4Hdr::LEN as i32 + CdHdr::LEN as i32);
 
-        Error::NoEntry(SupportedChannel::UdpToChan66)
-    })?;
+    adjust_head(ctx, NET_EXPANSION)?;
 
-    let new_src = ipv6.dst(); // The IP we received the packet on will be the new source IP.
-    let new_dst = client_and_channel.client_ip();
-    let new_ipv6_total_len = ipv6.payload_len() + CdHdr::LEN as u16;
+    // Now read the old packet data from its NEW location (shifted by 24 bytes)
+    let old_data_offset = -NET_EXPANSION as usize;
 
-    eth.swap_macs()?;
+    let (old_eth_src, old_eth_dst) = {
+        // SAFETY: This is the only mutable instance of `EthHdr` in this scope.
+        let old_eth = unsafe { ref_mut_at::<EthHdr>(ctx, old_data_offset)? };
+        (old_eth.src_addr, old_eth.dst_addr)
+    };
+
+    let (old_ipv4_src, old_ipv4_dst, old_ipv4_len, old_ipv4_tos, old_ipv4_ttl, old_ipv4_proto) = {
+        // SAFETY: This is the only mutable instance of `Ipv4Hdr` in this scope.
+        let old_ipv4 = unsafe { ref_mut_at::<Ipv4Hdr>(ctx, old_data_offset + EthHdr::LEN)? };
+        (
+            old_ipv4.src_addr(),
+            old_ipv4.dst_addr(),
+            old_ipv4.total_len(),
+            old_ipv4.tos,
+            old_ipv4.ttl,
+            old_ipv4.proto,
+        )
+    };
+
+    let (old_udp_len, old_udp_src, old_udp_dst, old_udp_check) = {
+        // SAFETY: This is the only mutable instance of `UdpHdr` in this scope.
+        let old_udp =
+            unsafe { ref_mut_at::<UdpHdr>(ctx, old_data_offset + EthHdr::LEN + Ipv4Hdr::LEN)? };
+        (
+            old_udp.len(),
+            old_udp.source(),
+            old_udp.dest(),
+            old_udp.check(),
+        )
+    };
 
-    let pseudo_header = ipv6.update(new_src, new_dst, new_ipv6_total_len);
+    //
+    // 1. Ethernet header
+    //
 
-    let udp_len = udp.len();
-    let new_udp_len = udp_len + CdHdr::LEN as u16;
+    // SAFETY: This is the only mutable instance of `EthHdr` in this scope.
+    let eth = unsafe { ref_mut_at::<EthHdr>(ctx, 0)? };
```

Comment:
Could put this into its own scope as well for consistency and easier verification of the safety?

What one actually has to verify is the offset you are passing. When the safety comment was written, we didn't have the offset parameter so the position in buffer was tied to the type and therefore, saying "it is the only type in the scope" was enough to guarantee thay there is no memory overlap.

Reply:
Is out of bounds the only possible UB here? I don't know unsafe code enough to say that for sure.

Reply:
Working on setting up a performance test harness tonight so we can see where the real bottlenecks lie. Won't block getting this out to prod though I think.